tests/network/vmi_istio.go

```diff
@@ -118,33 +120,67 @@ var _ = SIGDescribe("[Serial] Istio", func() {
 			tests.WaitUntilVMIReady(vmi, console.LoginToCirros)
 		})
 		Describe("Inbound traffic", func() {
-			table.DescribeTable("request to VMI", func(port int) {
-				job, err := createJobCheckingVMIReachability(vmi, port)
-				Expect(err).ShouldNot(HaveOccurred())
-
+			checkVMIReachability := func(vmi *v1.VirtualMachineInstance, targetPort int) error {
+				job, err := createJobCheckingVMIReachability(vmi, targetPort)
+				if err != nil {
+					return err
+				}
 				By("Waiting for the job to succeed")
-				Expect(tests.WaitForJobToSucceed(job, 480*time.Second)).To(Succeed())
-			},
-				table.Entry("should reach VMI HTTP server on service declared port", svcDeclaredTestPort),
-				table.Entry("should reach VMI HTTP server on service undeclared port", svcUndeclaredTestPort),
-			)
-			Context("With PeerAuthentication enforcing mTLS", func() {
+				return tests.WaitForJobToSucceed(job, 480*time.Second)
+			}
+
+			Context("With VMI having explicit ports specified", func() {
+				BeforeEach(func() {
+					vmiPorts = explicitPorts
+				})
```

Comment:
IIUC, the order of evaluation is : beforeEach, JustBeforeEach, test. 

I'd really appreciate if you'd define the `BeforeEach` clauses **before** the `JustBeforeEach` clauses - without that, I lose the ability of looking at the tests from top to bottom and get an ordered sense of events from it.

I personally think that using a pattern like 

```golang 
Context(func(){
    JustBeforeEach(func(){
        // do stuff
        // executed in t=2
    })
    Context(func(){
        BeforeEach(func(){
            // do stuff
            // executed in t=1
        })
        It("yadayada", func(){
           // do stuff
           // executed in t=3
        })
    })
})
```
should be discouraged.

Would it be possible for you to keep the temporal order of setup operations with the code (from top to bottom)? 

This makes it harder for me to reason about the code.

Reply:
> I'd really appreciate if you'd define the BeforeEach clauses before the JustBeforeEach clauses - without that, I lose the ability of looking at the tests from top to bottom and get an ordered sense of events from it.

I can't define the `JustBeforeEach` always after BeforeEach, because I want  `JustBeforeEach` to apply for all the tests.
At the same time, the configuration done in `BeforeEach` is supposed to be applied only in the specific context.

 > I personally think that using a pattern like
 This exact pattern is considered best practice in [ginkgo documentation](https://onsi.github.io/ginkgo/#separating-creation-and-configuration-justbeforeeach), which is why I use it.
I'm not saying that whatever is considered best practice in ginkgo docs should be followed in KubeVirt, but this seems like a general BDD concept to me.

It could get messy if abused for sure. In these tests however, I see the separation of the resource creation and configuration more helpful than hurting readability.

Reply:
I think `JustBeforeEach` is a nice tool that can help reduce duplications. However, I agree with Miguel it can confuse the reader.
In the case of this PR I would narrow the `JustBeforeEach` code only for the vm creation. Since AFAIU, the vm creation is  the only part that may have different "configs" between the tests.

Reply:
So the recommendation for use of `JustBeforeEach` on the network related tests is to limit its scope to whatever entities requiring it ? I.e. the VMI in this example. 

I think it's a fair compromise.

Reply:
> I think it's a fair compromise.

I agree.
I've removed all `JustBeforeEach` blocks except for the first one, which relies on configuration happening in `BeforeEach`.

Reply:
I don't follow; I see plenty of operations being performed:
   - label added to namespace
   - NAD created
   - VMI created
   - service create
   - wait for VMI creation

I was expecting the VMI being created & a blocking wait until it was ready.

Reply:
There is blocking wait here: https://github.com/kubevirt/kubevirt/pull/5461/commits/a196759da2131700f3ad049c2e556a3429bf7671#diff-549b5c85962b6a2029257ee04a44302196d99e8c66fe90d9b5032915f3731344R120

Reply:
I'm aware. I'm just pointing out that - iiuc - we've agreed this function to consist only of creating the VMI & waiting for it to be ready.

It does plenty more.

Reply:
The label and NAD is a requirement to create the VMI, I could move the service elsewhere, but it is closely related to the VMI, so IMO it makes sense to do it in the same place

Reply:
Well, moving the NAD creation to BeforeEach probably makes more sense.
I still think it's nicer to keep the service creation close to the VMI

Reply:
> Well, moving the NAD creation to BeforeEach probably makes more sense.
> I still think it's nicer to keep the service creation close to the VMI

Why ? Can you elaborate as to why it is nicer to create the service in the same block as the vmi ? 

I can understand creating the **only** resource that needs a parameterized configuration in an `JustBeforeEach` block. 

Having some things in that block that do not need any parameters, but other similar ones in a regular before each block is just confusing; it makes me wonder what's so special about entity X that it must be created in this block where parameterized things are created - e.g. why are you OK with creating the NAD in the `BeforeEach` block but insist on creating the service here ? I find that puzzling.

Reply:
> e.g. why are you OK with creating the NAD in the BeforeEach block but insist on creating the service here ? I find that puzzling.

My reasoning for that was the the NAD is a resource that can be used by another VMI that is created later in the tests and used as a server. But the service is tightly related to the this very VMI, so creating it next to it makes it IMO more obvious.

But I understand that it doesn't comply with the said rule of thumb to only use JustBeforeEach for resources

Reply:
[Done](https://github.com/kubevirt/kubevirt/compare/9f2a08c198389d4d6e95fccbb6232661dd30e82e..8cdde389bb10d82617d0b618c551fd9bc5f7fef8) 
The label needs to be added in the `JustBeforeEach` though as it needs to be added when the VMI is being created.

Reply:
Is there something wrong with adding it before ? 

Just asking out of curiosity.